.justfile

```diff
@@ -0,0 +1,171 @@
+
+repo := `git rev-parse --show-toplevel`
+profile := env_var_or_default('PROFILE', "dev")
+github-event-name := env_var_or_default('GITHUB_EVENT_NAME', "none")
+
+
+######################################################################
+## Helper to print a message when calling `just`
+######################################################################
+[private]
+@default:
+  echo "Usage: just <recipe>"
+  just list-repo-recipes
+  echo "For further information, run 'just --help'"
+
+# List recipes in this file and from the calling directory
+[no-cd]
+[private]
+@usage:
+  echo "Usage: just <recipe>"
+  just list-local-recipes
+  just list-repo-recipes
+  echo "For further information, run 'just --help'"
+
+[no-cd]
+[private]
+@list-local-recipes:
+  echo '\nAvailable recipes:'
+  just --list --unsorted --list-heading ''
+
+[private]
+@list-repo-recipes:
+  echo "\nRepository recipes:"
+  just --list --unsorted --list-heading ''
+
+
+######################################################################
+## Helper to run a global command
+######################################################################
+
+# Runs yarn in the repository root
+[private]
+yarn *args:
+  yarn {{args}}
+
+
+######################################################################
+## Helper to run a command on an environmental condition
+######################################################################
+
+# Runs the provided command if `PROFILE` starts with `"dev"`
+[private]
+[no-cd]
+in-dev +command:
+  #!/usr/bin/env bash
+  set -euo pipefail
+  if [[ {{ profile }} =~ dev.* ]]; then
+    echo "{{command}}" >&2
+    {{command}}
+  fi
+
+# Runs the provided command if in a pull request
+[private]
+[no-cd]
+in-pr +command:
+  #!/usr/bin/env bash
+  set -euo pipefail
+  if [[ {{ github-event-name }} = pull_request ]]; then
+    echo "{{command}}" >&2
+    {{command}}
+  fi
+
+# Runs the provided command if not in a pull request
+[private]
+[no-cd]
+not-in-pr +command:
+  #!/usr/bin/env bash
+  set -euo pipefail
+  if [[ {{ github-event-name }} != pull_request ]]; then
+    echo "{{command}}" >&2
+    {{command}}
+  fi
+
+
+######################################################################
+## Install scripts
+######################################################################
+
+[private]
+@install-cargo-tool tool install version:
+  `{{tool}} --version | grep -q "{{version}}" || cargo install "{{install}}" --version "{{version}}" --locked --force`
+
+[private]
+@install-cargo-hack:
+  just install-cargo-tool 'cargo hack' cargo-hack 0.5.26
+
+[private]
+@install-cargo-nextest:
+  just install-cargo-tool 'cargo nextest' cargo-nextest 0.9.37
+
+[private]
+@install-cargo-script:
+  just install-cargo-tool 'cargo script' cargo-script 0.2.8
+
+[private]
+@install-llvm-cov:
+  just install-cargo-tool 'cargo llvm-cov' cargo-llvm-cov 0.5.9
+
+[private]
+@install-cargo-insta:
+  just install-cargo-tool 'cargo insta' cargo-insta 1.18.2
```

Comment:
Any specific reason why `just` is invoked instead of just referring to the recipe directly? (applies to most recipes in this file)

Reply:
This way it's possible to overwrite a recipe. E.g. the `lint` recipe calls a few other recipes. If these are overwritten by other justfiles, those are called instead, so the recipes in these files remain usable.

Reply:
Which way round does the overriding work sorry? I don't quite understand what's being described.

Are you suggesting a local justfile could override `install-cargo-tool` for example, and then if _within that local context_ you call `just install-cargo-script` it would use your overridden one? Or it would use the one defined in this file?